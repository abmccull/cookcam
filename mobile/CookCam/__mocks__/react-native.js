const React = require('react');
const RN = jest.requireActual('react-native');

// Mock PixelRatio for StyleSheet
const PixelRatio = {
  roundToNearestPixel: (value) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size) => size * 2,
};

// Override modules that cause issues in tests
module.exports = {
  ...RN,
  PixelRatio,
  StyleSheet: {
    ...RN.StyleSheet,
    create: (styles) => styles,
    flatten: (style) => {
      if (!style) return {};
      if (Array.isArray(style)) {
        return Object.assign({}, ...style.filter(s => s));
      }
      return style;
    },
    absoluteFill: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    absoluteFillObject: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    hairlineWidth: 1,
  },
  Platform: {
    ...RN.Platform,
    OS: 'ios',
    select: (obj) => obj.ios || obj.default,
    Version: 14,
    isTV: false,
    isTesting: true,
  },
  Dimensions: {
    get: (dim) => {
      if (dim === 'window') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      if (dim === 'screen') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      return { width: 375, height: 812 };
    },
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    set: jest.fn(),
  },
  Animated: {
    ...RN.Animated,
    timing: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    spring: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    Value: class {
      constructor(value) {
        this._value = value;
      }
      setValue(value) {
        this._value = value;
      }
      interpolate(config) {
        return this;
      }
      addListener(callback) {
        return '1';
      }
      removeListener(id) {}
      removeAllListeners() {}
      stopAnimation(callback) {
        callback && callback(this._value);
      }
      resetAnimation(callback) {
        callback && callback(this._value);
      }
      setOffset(offset) {}
      flattenOffset() {}
      extractOffset() {}
    },
    View: RN.View,
    Text: RN.Text,
    Image: RN.Image,
    ScrollView: RN.ScrollView,
    createAnimatedComponent: (Component) => Component,
    parallel: (animations, config) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    sequence: (animations) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    loop: (animation) => ({
      start: (callback) => {
        animation.start && animation.start();
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    event: () => jest.fn(),
    add: jest.fn(),
    subtract: jest.fn(),
    divide: jest.fn(),
    multiply: jest.fn(),
    modulo: jest.fn(),
    diffClamp: jest.fn(),
  },
  NativeModules: {
    ...RN.NativeModules,
    UIManager: {
      RCTView: {
        directEventTypes: {},
      },
    },
    PlatformConstants: {
      forceTouchAvailable: false,
    },
    RNCNetInfo: {
      getCurrentState: jest.fn(() => Promise.resolve({
        isConnected: true,
        isInternetReachable: true,
        type: 'wifi'
      })),
      addListener: jest.fn(),
      removeListeners: jest.fn(),
    },
    StatusBarManager: {
      HEIGHT: 20,
      getHeight: jest.fn(),
      setColor: jest.fn(),
      setStyle: jest.fn(),
      setHidden: jest.fn(),
      setNetworkActivityIndicatorVisible: jest.fn(),
      setBackgroundColor: jest.fn(),
      setTranslucent: jest.fn(),
    },
    SettingsManager: {
      settings: {},
      setValues: jest.fn(),
      deleteValues: jest.fn(),
    },
    DeviceInfo: {
      Dimensions: {
        window: { width: 375, height: 812, scale: 2, fontScale: 1 },
        screen: { width: 375, height: 812, scale: 2, fontScale: 1 },
      },
    },
  },
  TouchableOpacity: React.forwardRef((props, ref) =>
    React.createElement('TouchableOpacity', { ...props, ref })
  ),
  TouchableHighlight: React.forwardRef((props, ref) =>
    React.createElement('TouchableHighlight', { ...props, ref })
  ),
  TouchableWithoutFeedback: React.forwardRef((props, ref) =>
    React.createElement('TouchableWithoutFeedback', { ...props, ref })
  ),
  TouchableNativeFeedback: {
    SelectableBackground: jest.fn(),
    SelectableBackgroundBorderless: jest.fn(),
    Ripple: jest.fn(),
    canUseNativeForeground: jest.fn(),
  },
  View: React.forwardRef((props, ref) =>
    React.createElement('View', { ...props, ref })
  ),
  Text: React.forwardRef((props, ref) =>
    React.createElement('Text', { ...props, ref })
  ),
  Image: React.forwardRef((props, ref) =>
    React.createElement('Image', { ...props, ref })
  ),
  ScrollView: React.forwardRef((props, ref) =>
    React.createElement('ScrollView', { ...props, ref })
  ),
  TextInput: React.forwardRef((props, ref) =>
    React.createElement('TextInput', { ...props, ref })
  ),
  Modal: React.forwardRef((props, ref) =>
    React.createElement('Modal', { ...props, ref })
  ),
  FlatList: React.forwardRef((props, ref) =>
    React.createElement('FlatList', { ...props, ref })
  ),
  SectionList: React.forwardRef((props, ref) =>
    React.createElement('SectionList', { ...props, ref })
  ),
  VirtualizedList: React.forwardRef((props, ref) =>
    React.createElement('VirtualizedList', { ...props, ref })
  ),
  Alert: {
    alert: jest.fn(),
    prompt: jest.fn(),
  },
  Keyboard: {
    addListener: jest.fn(() => ({ remove: jest.fn() })),
    removeListener: jest.fn(),
    dismiss: jest.fn(),
    scheduleLayoutAnimation: jest.fn(),
  },
  Linking: {
    openURL: jest.fn(() => Promise.resolve()),
    canOpenURL: jest.fn(() => Promise.resolve(true)),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AppState: {
    currentState: 'active',
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AccessibilityInfo: {
    isScreenReaderEnabled: jest.fn(() => Promise.resolve(false)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    announceForAccessibility: jest.fn(),
    setAccessibilityFocus: jest.fn(),
  },
  findNodeHandle: jest.fn(() => 1),
  unstable_batchedUpdates: (callback) => callback(),
};