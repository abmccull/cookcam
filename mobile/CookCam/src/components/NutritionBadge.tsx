import React, { useMemo } from "react";
import { View, Text, StyleSheet } from "react-native";
import { Zap, Activity } from "lucide-react-native";
import {
  scale,
  verticalScale,
  moderateScale,
  responsive,
} from "../utils/responsive";

interface NutritionData {
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g?: number;
  sodium_mg?: number;
}

interface NutritionBadgeProps {
  nutrition: NutritionData;
  servings?: number;
  variant?: "full" | "compact";
}

const NutritionBadge: React.FC<NutritionBadgeProps> = React.memo(
  ({ nutrition, servings = 1, variant = "full" }) => {
    const isCompact = variant === "compact";

    // Memoize nutrition calculations
    const nutritionPerServing = useMemo(
      () => ({
        calories: Math.round(nutrition.calories / servings),
        protein: Math.round(nutrition.protein_g / servings),
        carbs: Math.round(nutrition.carbs_g / servings),
        fat: Math.round(nutrition.fat_g / servings),
        fiber: nutrition.fiber_g
          ? Math.round(nutrition.fiber_g / servings)
          : undefined,
        sodium: nutrition.sodium_mg
          ? Math.round(nutrition.sodium_mg / servings)
          : undefined,
      }),
      [nutrition, servings],
    );

    // Memoize calorie level classification
    const calorieLevel = useMemo(() => {
      const calories = nutritionPerServing.calories;
      if (calories < 200) return { level: "low", color: "#4CAF50" };
      if (calories < 400) return { level: "moderate", color: "#FF9800" };
      return { level: "high", color: "#F44336" };
    }, [nutritionPerServing.calories]);

    // Memoize protein level classification
    const proteinLevel = useMemo(() => {
      const protein = nutritionPerServing.protein;
      if (protein >= 20) return { level: "high", color: "#4CAF50" };
      if (protein >= 10) return { level: "moderate", color: "#FF9800" };
      return { level: "low", color: "#9E9E9E" };
    }, [nutritionPerServing.protein]);

    if (isCompact) {
      return (
        <View style={styles.compactContainer}>
          <View
            style={[
              styles.compactBadge,
              { backgroundColor: calorieLevel.color },
            ]}
          >
            <Zap size={moderateScale(12)} color="#FFFFFF" />
            <Text style={styles.compactText}>
              {nutritionPerServing.calories}
            </Text>
          </View>
          <View
            style={[
              styles.compactBadge,
              { backgroundColor: proteinLevel.color },
            ]}
          >
            <Activity size={moderateScale(12)} color="#FFFFFF" />
            <Text style={styles.compactText}>
              {nutritionPerServing.protein}g
            </Text>
          </View>
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <View style={styles.header}>
          <Text style={styles.title}>Nutrition Facts</Text>
          <Text style={styles.servingText}>Per serving</Text>
        </View>

        <View style={styles.mainNutrient}>
          <View style={styles.calorieRow}>
            <Zap size={moderateScale(20)} color={calorieLevel.color} />
            <Text style={styles.calorieValue}>
              {nutritionPerServing.calories}
            </Text>
            <Text style={styles.calorieLabel}>calories</Text>
          </View>
        </View>

        <View style={styles.macroNutrients}>
          <View style={styles.macroItem}>
            <Activity size={moderateScale(16)} color={proteinLevel.color} />
            <Text style={styles.macroValue}>
              {nutritionPerServing.protein}g
            </Text>
            <Text style={styles.macroLabel}>protein</Text>
          </View>

          <View style={styles.macroItem}>
            <Text style={styles.macroValue}>{nutritionPerServing.carbs}g</Text>
            <Text style={styles.macroLabel}>carbs</Text>
          </View>

          <View style={styles.macroItem}>
            <Text style={styles.macroValue}>{nutritionPerServing.fat}g</Text>
            <Text style={styles.macroLabel}>fat</Text>
          </View>
        </View>

        {(nutritionPerServing.fiber || nutritionPerServing.sodium) && (
          <View style={styles.additionalNutrients}>
            {nutritionPerServing.fiber && (
              <Text style={styles.additionalText}>
                Fiber: {nutritionPerServing.fiber}g
              </Text>
            )}
            {nutritionPerServing.sodium && (
              <Text style={styles.additionalText}>
                Sodium: {nutritionPerServing.sodium}mg
              </Text>
            )}
          </View>
        )}
      </View>
    );
  },
);

// Add display name for debugging
NutritionBadge.displayName = "NutritionBadge";

const styles = StyleSheet.create({
  compactContainer: {
    flexDirection: "row",
    alignItems: "center",
  },
  compactBadge: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: scale(8),
    paddingVertical: scale(4),
    borderRadius: responsive.borderRadius.medium,
    gap: scale(3),
  },
  compactText: {
    fontSize: responsive.fontSize.small,
    fontWeight: "700",
    color: "#FFFFFF",
  },
  container: {
    backgroundColor: "#F8F9FA",
    borderRadius: responsive.borderRadius.medium,
    padding: scale(12),
    marginTop: verticalScale(8),
  },
  header: {
    flexDirection: "row",
    alignItems: "center",
    gap: scale(6),
    marginBottom: verticalScale(8),
  },
  title: {
    fontSize: responsive.fontSize.regular,
    fontWeight: "600",
    color: "#4CAF50",
  },
  servingText: {
    fontSize: responsive.fontSize.tiny,
    fontWeight: "500",
    color: "#666",
  },
  mainNutrient: {
    marginBottom: verticalScale(16),
  },
  calorieRow: {
    flexDirection: "row",
    alignItems: "center",
    gap: scale(6),
  },
  calorieValue: {
    fontSize: responsive.fontSize.regular,
    fontWeight: "700",
    color: "#2D1B69",
  },
  calorieLabel: {
    fontSize: responsive.fontSize.tiny,
    fontWeight: "500",
    color: "#666",
    marginTop: verticalScale(2),
  },
  macroNutrients: {
    flexDirection: "row",
    justifyContent: "space-between",
  },
  macroItem: {
    alignItems: "center",
    flex: 1,
  },
  macroValue: {
    fontSize: responsive.fontSize.regular,
    fontWeight: "700",
    color: "#2D1B69",
  },
  macroLabel: {
    fontSize: responsive.fontSize.tiny,
    fontWeight: "500",
    color: "#666",
    marginTop: verticalScale(2),
  },
  additionalNutrients: {
    marginTop: verticalScale(16),
  },
  additionalText: {
    fontSize: responsive.fontSize.tiny,
    fontWeight: "500",
    color: "#666",
  },
});

export default NutritionBadge;
