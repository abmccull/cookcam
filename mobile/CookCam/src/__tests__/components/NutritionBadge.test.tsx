// Set up mocks before imports
global.PixelRatio = {
  roundToNearestPixel: (value) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size) => size * 2,
};

// Mock React Native
jest.mock('react-native', () => {
  const actualRN = jest.requireActual('react-native');
  actualRN.PixelRatio = global.PixelRatio;
  return {
    ...actualRN,
    PixelRatio: global.PixelRatio,
    StyleSheet: {
      create: (styles) => styles,
      flatten: (style) => style,
    },
  };
});

// Mock environment config
jest.mock('../../config/env', () => ({
  __esModule: true,
  default: () => ({
    SUPABASE_URL: "https://test.supabase.co",
    SUPABASE_ANON_KEY: "test-anon-key",
    API_BASE_URL: "https://test-api.cookcam.com",
  }),
}));

// Mock responsive utilities  
jest.mock('../../utils/responsive', () => ({
  scale: (size) => size,
  verticalScale: (size) => size,
  moderateScale: (size) => size,
  responsive: {
    fontSize: {
      tiny: 10,
      small: 12,
      regular: 14,
      large: 16,
    },
    borderRadius: {
      small: 4,
      medium: 8,
      large: 12,
    },
  },
}));

// Mock icons
jest.mock('lucide-react-native', () => ({
  Zap: 'Zap',
  Activity: 'Activity',
}));

import React from 'react';
import { render, screen } from '@testing-library/react-native';
import NutritionBadge from '../../components/NutritionBadge';

describe('NutritionBadge', () => {
  const mockNutrition = {
    calories: 250,
    protein_g: 15,
    carbs_g: 30,
    fat_g: 10,
    fiber_g: 5,
    sodium_mg: 500,
  };

  const lowCalorieNutrition = {
    calories: 150,
    protein_g: 8,
    carbs_g: 20,
    fat_g: 5,
  };

  const highCalorieNutrition = {
    calories: 450,
    protein_g: 25,
    carbs_g: 50,
    fat_g: 20,
  };

  const highProteinNutrition = {
    calories: 300,
    protein_g: 30,
    carbs_g: 25,
    fat_g: 12,
  };

  describe('Full Variant', () => {
    it('should render correctly with all props', () => {
      const { toJSON } = render(<NutritionBadge nutrition={mockNutrition} />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render nutrition facts title', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('Nutrition Facts')).toBeTruthy();
      expect(screen.getByText('Per serving')).toBeTruthy();
    });

    it('should display calories', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('250')).toBeTruthy();
      expect(screen.getByText('calories')).toBeTruthy();
    });

    it('should display macronutrients', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('protein')).toBeTruthy();
      expect(screen.getByText('15g')).toBeTruthy();
      
      expect(screen.getByText('carbs')).toBeTruthy();
      expect(screen.getByText('30g')).toBeTruthy();
      
      expect(screen.getByText('fat')).toBeTruthy();
      expect(screen.getByText('10g')).toBeTruthy();
    });

    it('should display optional nutrients when provided', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('Fiber: 5g')).toBeTruthy();
      expect(screen.getByText('Sodium: 500mg')).toBeTruthy();
    });

    it('should not display optional nutrients when not provided', () => {
      const nutritionWithoutOptional = {
        calories: 250,
        protein_g: 15,
        carbs_g: 30,
        fat_g: 10,
      };
      
      render(<NutritionBadge nutrition={nutritionWithoutOptional} />);
      
      expect(screen.queryByText(/Fiber:/)).toBeFalsy();
      expect(screen.queryByText(/Sodium:/)).toBeFalsy();
    });

    it('should display icons in full mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="full" />);
      
      // Zap icon for calories
      expect(screen.UNSAFE_getAllByType('Zap')).toHaveLength(1);
      // Activity icon for protein
      expect(screen.UNSAFE_getAllByType('Activity')).toHaveLength(1);
    });
  });

  describe('Compact Variant', () => {
    it('should render correctly in compact mode', () => {
      const { toJSON } = render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render compact badges', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      // Should have two badges
      expect(screen.UNSAFE_getAllByType('View').length).toBeGreaterThan(1);
    });

    it('should display calories in compact mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      expect(screen.getByText('250')).toBeTruthy();
    });

    it('should display protein in compact mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      expect(screen.getByText('15g')).toBeTruthy();
    });

    it('should not display full nutrition facts in compact mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      expect(screen.queryByText('Nutrition Facts')).toBeFalsy();
      expect(screen.queryByText('carbs')).toBeFalsy();
      expect(screen.queryByText('fat')).toBeFalsy();
      expect(screen.queryByText('Per serving')).toBeFalsy();
    });

    it('should display icons in compact mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      // Zap icon for calories
      expect(screen.UNSAFE_getAllByType('Zap')).toHaveLength(1);
      // Activity icon for protein
      expect(screen.UNSAFE_getAllByType('Activity')).toHaveLength(1);
    });
  });

  describe('Servings Calculation', () => {
    it('should calculate per serving for 2 servings', () => {
      render(<NutritionBadge nutrition={mockNutrition} servings={2} />);
      
      expect(screen.getByText('125')).toBeTruthy(); // 250/2
      expect(screen.getByText('8g')).toBeTruthy(); // 15/2 rounded
    });

    it('should calculate per serving for 4 servings', () => {
      render(<NutritionBadge nutrition={mockNutrition} servings={4} />);
      
      expect(screen.getByText('63')).toBeTruthy(); // 250/4 rounded
      expect(screen.getByText('4g')).toBeTruthy(); // 15/4 rounded
    });

    it('should default to 1 serving when not specified', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('250')).toBeTruthy();
      expect(screen.getByText('15g')).toBeTruthy();
    });

    it('should round values correctly', () => {
      const nutrition = {
        calories: 333,
        protein_g: 17,
        carbs_g: 41,
        fat_g: 13,
      };
      
      render(<NutritionBadge nutrition={nutrition} servings={2} />);
      
      expect(screen.getByText('167')).toBeTruthy(); // 333/2 rounded
      expect(screen.getByText('9g')).toBeTruthy(); // 17/2 rounded
    });
  });

  describe('Calorie Level Classification', () => {
    it('should classify low calories (< 200)', () => {
      const { getByText } = render(<NutritionBadge nutrition={lowCalorieNutrition} variant="compact" />);
      
      // Low calorie should use green color (#4CAF50)
      const calorieText = getByText('150');
      expect(calorieText).toBeTruthy();
    });

    it('should classify moderate calories (200-400)', () => {
      const { getByText } = render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      // Moderate calorie should use orange color (#FF9800)
      const calorieText = getByText('250');
      expect(calorieText).toBeTruthy();
    });

    it('should classify high calories (> 400)', () => {
      const { getByText } = render(<NutritionBadge nutrition={highCalorieNutrition} variant="compact" />);
      
      // High calorie should use red color (#F44336)
      const calorieText = getByText('450');
      expect(calorieText).toBeTruthy();
    });
  });

  describe('Protein Level Classification', () => {
    it('should classify low protein (< 10g)', () => {
      render(<NutritionBadge nutrition={lowCalorieNutrition} variant="compact" />);
      
      expect(screen.getByText('8g')).toBeTruthy();
    });

    it('should classify moderate protein (10-20g)', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      expect(screen.getByText('15g')).toBeTruthy();
    });

    it('should classify high protein (>= 20g)', () => {
      render(<NutritionBadge nutrition={highProteinNutrition} variant="compact" />);
      
      expect(screen.getByText('30g')).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle zero values', () => {
      const zeroNutrition = {
        calories: 0,
        protein_g: 0,
        carbs_g: 0,
        fat_g: 0,
      };
      
      render(<NutritionBadge nutrition={zeroNutrition} />);
      
      expect(screen.getByText('0')).toBeTruthy();
      expect(screen.getAllByText('0g')).toHaveLength(3); // protein, carbs, fat
    });

    it('should handle very large values', () => {
      const largeNutrition = {
        calories: 9999,
        protein_g: 999,
        carbs_g: 999,
        fat_g: 999,
      };
      
      render(<NutritionBadge nutrition={largeNutrition} />);
      
      expect(screen.getByText('9999')).toBeTruthy();
      expect(screen.getAllByText('999g')).toHaveLength(3);
    });

    it('should handle decimal values with rounding', () => {
      const decimalNutrition = {
        calories: 255.7,
        protein_g: 15.4,
        carbs_g: 30.8,
        fat_g: 10.2,
      };
      
      render(<NutritionBadge nutrition={decimalNutrition} />);
      
      // Values should be rounded
      expect(screen.getByText('256')).toBeTruthy();
      expect(screen.getByText('15g')).toBeTruthy();
      expect(screen.getByText('31g')).toBeTruthy();
      expect(screen.getByText('10g')).toBeTruthy();
    });

    it('should handle negative values', () => {
      const negativeNutrition = {
        calories: -100,
        protein_g: -5,
        carbs_g: -10,
        fat_g: -3,
      };
      
      render(<NutritionBadge nutrition={negativeNutrition} />);
      
      // Should still render, even with negative values
      expect(screen.getByText('-100')).toBeTruthy();
    });
  });

  describe('Memoization', () => {
    it('should have display name for debugging', () => {
      expect(NutritionBadge.displayName).toBe('NutritionBadge');
    });

    it('should memoize nutrition calculations', () => {
      const { rerender } = render(<NutritionBadge nutrition={mockNutrition} servings={2} />);
      
      // Initial render
      expect(screen.getByText('125')).toBeTruthy();
      
      // Re-render with same props
      rerender(<NutritionBadge nutrition={mockNutrition} servings={2} />);
      
      // Should still show same values
      expect(screen.getByText('125')).toBeTruthy();
    });

    it('should recalculate when servings change', () => {
      const { rerender } = render(<NutritionBadge nutrition={mockNutrition} servings={2} />);
      
      expect(screen.getByText('125')).toBeTruthy();
      
      rerender(<NutritionBadge nutrition={mockNutrition} servings={4} />);
      
      expect(screen.getByText('63')).toBeTruthy();
    });

    it('should recalculate when nutrition changes', () => {
      const { rerender } = render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('250')).toBeTruthy();
      
      rerender(<NutritionBadge nutrition={highCalorieNutrition} />);
      
      expect(screen.getByText('450')).toBeTruthy();
    });

    it('should maintain variant when re-rendering', () => {
      const { rerender } = render(<NutritionBadge nutrition={mockNutrition} variant="compact" />);
      
      expect(screen.queryByText('Nutrition Facts')).toBeFalsy();
      
      rerender(<NutritionBadge nutrition={mockNutrition} variant="full" />);
      
      expect(screen.getByText('Nutrition Facts')).toBeTruthy();
    });
  });

  describe('Display Text', () => {
    it('should use lowercase labels in full mode', () => {
      render(<NutritionBadge nutrition={mockNutrition} variant="full" />);
      
      expect(screen.getByText('calories')).toBeTruthy();
      expect(screen.getByText('protein')).toBeTruthy();
      expect(screen.getByText('carbs')).toBeTruthy();
      expect(screen.getByText('fat')).toBeTruthy();
    });

    it('should format additional nutrients correctly', () => {
      render(<NutritionBadge nutrition={mockNutrition} />);
      
      expect(screen.getByText('Fiber: 5g')).toBeTruthy();
      expect(screen.getByText('Sodium: 500mg')).toBeTruthy();
    });

    it('should not show additional nutrients section when both are missing', () => {
      const nutritionWithoutOptional = {
        calories: 250,
        protein_g: 15,
        carbs_g: 30,
        fat_g: 10,
      };
      
      render(<NutritionBadge nutrition={nutritionWithoutOptional} />);
      
      // Should not have the additional nutrients section at all
      expect(screen.queryByText(/Fiber:/)).toBeFalsy();
      expect(screen.queryByText(/Sodium:/)).toBeFalsy();
    });

    it('should show only fiber when sodium is missing', () => {
      const nutritionWithFiber = {
        calories: 250,
        protein_g: 15,
        carbs_g: 30,
        fat_g: 10,
        fiber_g: 8,
      };
      
      render(<NutritionBadge nutrition={nutritionWithFiber} />);
      
      expect(screen.getByText('Fiber: 8g')).toBeTruthy();
      expect(screen.queryByText(/Sodium:/)).toBeFalsy();
    });

    it('should show only sodium when fiber is missing', () => {
      const nutritionWithSodium = {
        calories: 250,
        protein_g: 15,
        carbs_g: 30,
        fat_g: 10,
        sodium_mg: 750,
      };
      
      render(<NutritionBadge nutrition={nutritionWithSodium} />);
      
      expect(screen.getByText('Sodium: 750mg')).toBeTruthy();
      expect(screen.queryByText(/Fiber:/)).toBeFalsy();
    });
  });
});