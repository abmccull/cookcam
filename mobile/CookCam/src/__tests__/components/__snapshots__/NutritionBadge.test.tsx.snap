// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NutritionBadge Compact Variant should render correctly in compact mode 1`] = `
<View
  ref={null}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    ref={null}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 3,
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        {
          "backgroundColor": "#FF9800",
        },
      ]
    }
  >
    <Zap
      color="#FFFFFF"
      size={12}
    />
    <Text
      ref={null}
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "fontWeight": "700",
        }
      }
    >
      250
    </Text>
  </View>
  <View
    ref={null}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "gap": 3,
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        {
          "backgroundColor": "#FF9800",
        },
      ]
    }
  >
    <Activity
      color="#FFFFFF"
      size={12}
    />
    <Text
      ref={null}
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "fontWeight": "700",
        }
      }
    >
      15
      g
    </Text>
  </View>
</View>
`;

exports[`NutritionBadge Full Variant should render correctly with all props 1`] = `
<View
  ref={null}
  style={
    {
      "backgroundColor": "#F8F9FA",
      "borderRadius": 8,
      "marginTop": 8,
      "padding": 12,
    }
  }
>
  <View
    ref={null}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 6,
        "marginBottom": 8,
      }
    }
  >
    <Text
      ref={null}
      style={
        {
          "color": "#4CAF50",
          "fontSize": 14,
          "fontWeight": "600",
        }
      }
    >
      Nutrition Facts
    </Text>
    <Text
      ref={null}
      style={
        {
          "color": "#666",
          "fontSize": 10,
          "fontWeight": "500",
        }
      }
    >
      Per serving
    </Text>
  </View>
  <View
    ref={null}
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      ref={null}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 6,
        }
      }
    >
      <Zap
        color="#FF9800"
        size={20}
      />
      <Text
        ref={null}
        style={
          {
            "color": "#2D1B69",
            "fontSize": 14,
            "fontWeight": "700",
          }
        }
      >
        250
      </Text>
      <Text
        ref={null}
        style={
          {
            "color": "#666",
            "fontSize": 10,
            "fontWeight": "500",
            "marginTop": 2,
          }
        }
      >
        calories
      </Text>
    </View>
  </View>
  <View
    ref={null}
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      ref={null}
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Activity
        color="#FF9800"
        size={16}
      />
      <Text
        ref={null}
        style={
          {
            "color": "#2D1B69",
            "fontSize": 14,
            "fontWeight": "700",
          }
        }
      >
        15
        g
      </Text>
      <Text
        ref={null}
        style={
          {
            "color": "#666",
            "fontSize": 10,
            "fontWeight": "500",
            "marginTop": 2,
          }
        }
      >
        protein
      </Text>
    </View>
    <View
      ref={null}
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        ref={null}
        style={
          {
            "color": "#2D1B69",
            "fontSize": 14,
            "fontWeight": "700",
          }
        }
      >
        30
        g
      </Text>
      <Text
        ref={null}
        style={
          {
            "color": "#666",
            "fontSize": 10,
            "fontWeight": "500",
            "marginTop": 2,
          }
        }
      >
        carbs
      </Text>
    </View>
    <View
      ref={null}
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        ref={null}
        style={
          {
            "color": "#2D1B69",
            "fontSize": 14,
            "fontWeight": "700",
          }
        }
      >
        10
        g
      </Text>
      <Text
        ref={null}
        style={
          {
            "color": "#666",
            "fontSize": 10,
            "fontWeight": "500",
            "marginTop": 2,
          }
        }
      >
        fat
      </Text>
    </View>
  </View>
  <View
    ref={null}
    style={
      {
        "marginTop": 16,
      }
    }
  >
    <Text
      ref={null}
      style={
        {
          "color": "#666",
          "fontSize": 10,
          "fontWeight": "500",
        }
      }
    >
      Fiber: 
      5
      g
    </Text>
    <Text
      ref={null}
      style={
        {
          "color": "#666",
          "fontSize": 10,
          "fontWeight": "500",
        }
      }
    >
      Sodium: 
      500
      mg
    </Text>
  </View>
</View>
`;
